Setup e2e tests for webui

bafonins opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

References https://github.com/TheThingsNetwork/lorawan-stack/issues/2605


#### Changes
<!-- What are the changes made in this pull request? -->

- Extend documentation with the e2e testing methodology.
- Test coverage for e2e tests.
- Add base cypress tests with base utilities.


#### Testing

<!-- How did you verify that this change works? -->

See the docs


#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

~There are no big changes to the magefile compared to https://github.com/TheThingsNetwork/lorawan-stack/pull/2764. I only fixed some minor formatting issues. Will implement the rest after discussing the tests.~

@kschiffer 
~At this stage of this PR please review the proposed testing methodology and the tests. Ill follow up with changes to mage later. Let me know if there is anything you dont agree with or want to be clarified.~

<details>
<summary>Details</summary>

1. **Changing stack configuration.** As mentioned in the docs, UI for some views can vary depending on the stack configuration. Currently, there is no straightforward way to change the configuration between the tests except for restarting the lorawan-stack with different flags. This takes quite some time and drastically complicates the setup for e2e tests. So, as of now, we use the default configuration (with all components enabled), but edit only the injected `<script />` tag when necessary. IMO, this should suffice as it is the only source of truth for the frontend to determine the stack configuration. See tests for creating applications. Ideally, we should test what we actually deploy, but currently this does not seem feasible.
2. **Asserting on backend related messages.** Overall, I dont think that doing this is a good idea. I dont like the idea of making assertions of on specific text captions that the frontend has no control of. Such approach might force us to adjust the tests every time when, for exampple, validation order on the backend changes. However, in some specific cases this can be helpful.
3. **Asserting on API request body, redux store.** I dont really have any strong opinions on this and I can imagine that for some cases making assertions on api responses/requests can be helpful.
4. **Title assertions.** Sometimes these fail in interactive mode. Does this happen to you as well @kschiffer ?


</details>

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+1. `pkg/cypress/integrations` contains all test specifications.
+  - Each test file must be placed into corresponding folder (`console`/`oauth`).
+  - Each test must follow the following naming: `{context}.spec.js`.
+  - One test file must have end-to-end dedicated only to a specific entity or view.
```
```suggestion
  - One test file must have end-to-end tests dedicated only to a specific entity or view.
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+- Do not repeat actions via the UI that are not related to the current test context. Consider adding reusable [Cypress commands](https://docs.cypress.io/api/cypress-api/custom-commands.html) that do necessary test setup programmatically. This means that when testing any UI that is not the login specification and requies the user to be logged in, there is no need to log in through the login page,  ...
+- Extract components that appear on various views and test them separately instead of making assertions in each test where this component is used. For example, such components could be the page header and side navigation.
+- Dedicate at least one test to assert that the view displays its UI elements in place on initial load. Assert on UI changes in tests that trigger these changes.
+- Prefer duplicating entities with non-conflicting ids in tests instead of executing database teardown before each test. For example, when testing various scenaros for registering gateway, consider creating gateways with different ID's and EUI's instead of using a single gateway and drop the database before each test. Note: try to use this approach when possible, otherwise do not hesitate to rest ...
```
```suggestion
- Prefer duplicating entities with non-conflicting ids in tests instead of executing database teardown before each test. For example, when testing various scenarios for registering gateway, consider creating gateways with different ID's and EUI's instead of using a single gateway and drop the database before each test. Note: try to use this approach when possible, otherwise do not hesitate to restore database state before each test.
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+- Extract components that appear on various views and test them separately instead of making assertions in each test where this component is used. For example, such components could be the page header and side navigation.
+- Dedicate at least one test to assert that the view displays its UI elements in place on initial load. Assert on UI changes in tests that trigger these changes.
+- Prefer duplicating entities with non-conflicting ids in tests instead of executing database teardown before each test. For example, when testing various scenaros for registering gateway, consider creating gateways with different ID's and EUI's instead of using a single gateway and drop the database before each test. Note: try to use this approach when possible, otherwise do not hesitate to rest ...
+- Consider various stack configurations when writing end-to-end tests. Some views have different UI depending on availability of diffenent stack components. For example, the end device wizard looks different for deployments with complete cluster (NS+JS+AS) and for JS-only configuration.
```
```suggestion
- Consider various stack configurations when writing end-to-end tests. Some views have different UI depending on availability of diffenent stack components. For example, the end device wizard looks different for deployments with complete cluster (NS+JS+AS) and for JS-only configuration. Likewise, sections and entire views can be enabled or disabled based on our feature toggles. If your test scenario differs based on different feature toggle conditions, make sure to probe these preconditions in your tests.
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+- Prefer duplicating entities with non-conflicting ids in tests instead of executing database teardown before each test. For example, when testing various scenaros for registering gateway, consider creating gateways with different ID's and EUI's instead of using a single gateway and drop the database before each test. Note: try to use this approach when possible, otherwise do not hesitate to rest ...
+- Consider various stack configurations when writing end-to-end tests. Some views have different UI depending on availability of diffenent stack components. For example, the end device wizard looks different for deployments with complete cluster (NS+JS+AS) and for JS-only configuration.
+
+We distinguish between regular end-to-end tests and smoke tests. With smoke tests we try to accumulate application state (redux store, local storage, application cache, etc.) and verify that the application works as expected. When writing smoke tests we comply with he following guidelines:
```
```suggestion
##### Smoke tests

We distinguish between regular end-to-end tests and smoke tests. With smoke tests we try to accumulate application state (redux store, local storage, application cache, etc.) and verify that the application works as expected. When writing smoke tests we comply with the following guidelines:
```

kschiffer commented on cypress/integration/oauth/change-password.spec.js:
```diff
@@ -0,0 +1,115 @@
+    cy.findByLabelText('Old password').type(user.password)
+    cy.findByLabelText('New password').type(newPassword)
+    cy.findByLabelText('Confirm password').type(`${newPassword}{enter}`)
+
```
```suggestion
cy.findByLabelText('Revoke access').check()
```

To keep this working regardless of the default state.

kschiffer commented on cypress/integration/oauth/login.spec.js:
```diff
@@ -0,0 +1,88 @@
+      .and('be.visible')
+    cy.get('@userIdError')
+      .should('contain', 'is required')
+      .and('be.visible')
```
```suggestion
    cy.get('@userIdError')
      .should('contain.text', 'User ID is required')
      .and('be.visible')
```

Same for all similar occurrences.

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+
+We distinguish between regular end-to-end tests and smoke tests. With smoke tests we try to accumulate application state (redux store, local storage, application cache, etc.) and verify that the application works as expected. When writing smoke tests we comply with he following guidelines:
+
+- Smoke tests are testing complete user behavior to perform some complex and critical flow. For example, when testing registration of The Things Uno:
```
```suggestion
- Smoke tests are testing complete user stories in a **wide and shallow** manner, meaning:
  - performing some complex and critical flow that touches multiple components, APIs and/or views
  - not testing different configurations or preconditions of the same flow in depth
- For example, when testing registration of The Things Uno:
 
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+- Smoke tests are testing complete user behavior to perform some complex and critical flow. For example, when testing registration of The Things Uno:
+  1. Add test file `cypress/integration/smoke/devices/create.js`
+  2. Describe the whole user story to register the device including logging in, creating an application (or using an existing one), link the application and create the end device.
+- One smoke test should be encapsulated into a single `it` declaration.
```
This is a bit confusing now since you use a helper function `describeSmokeTest()`

kschiffer commented on cypress/integration/smoke/utils.js:
```diff
@@ -0,0 +1,22 @@
+
+/* eslint-disable import/prefer-default-export */
+
+const defineSmokeTest = (description, test) => ({
+  description,
+  test,
+})
```
I understand the idea but I'd rather be able to use the native `describe() / it()` syntax for the smoke tests. E.g. we might want to group related smoke tests using `describe()`.

bafonins replied:
I added `defineSmokeTest` to integrate other features to smoke tests later on. Please read https://cypress.io/blog/2019/12/06/use-meaningful-smoke-tests/

kschiffer commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,170 @@
+})
+
+// Helper function to quickly login to the console programmatically.
+Cypress.Commands.add('loginConsole', credentials => {
```
```suggestion
Cypress.Commands.add('loginConsole', credentials => {
```

Let's use a default argument for the credentials. For most test cases it should not be necessary to use distinct credentials.

bafonins replied:
See https://github.com/TheThingsNetwork/lorawan-stack/pull/2939#issuecomment-662472814

kschiffer commented on pkg/webui/template.go:
```diff
@@ -89,17 +89,26 @@ const appHTML = `
-			window.SENTRY_DSN={{.SentryDSN}};
-			{{with .CSRFToken}}window.CSRF_TOKEN={{.}};{{end}}
-      {{with .PageData}}window.PAGE_DATA={{.}};{{end}}
+		<script>
```
```suggestion
	<script>
```

kschiffer commented on pkg/webui/template.go:
```diff
@@ -89,17 +89,26 @@ const appHTML = `
+					SITE_SUB_TITLE:{{.SubTitle}},
+					SENTRY_DSN:{{.SentryDSN}},
+					{{with .CSRFToken}}CSRF_TOKEN:{{.}}{{end}},
+					{{with .PageData}}PAGE_DATA:{{.}}{{end}}
```
Notice the ending comma in the conditional, I think this will result in errors in some situations.

bafonins replied:
Not sure what you mean here. Can you provide code suggestion?

kschiffer commented on pkg/webui/template.go:
```diff
@@ -89,17 +89,26 @@ const appHTML = `
+				};
+		  win.__ttn_config__ = config;
+
+			if (win.Cypress && win.__initStackConfig) {
```
The tabs seem to be off here.

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,161 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+  2. Test the behavior of the API key create view independently from any other specification.
+- Do not repeat actions via the UI that are not related to the current test context. Consider adding reusable [Cypress commands](https://docs.cypress.io/api/cypress-api/custom-commands.html) that do necessary test setup programmatically. This means that when testing any UI that is not the login specification and requies the user to be logged in, there is no need to log in through the login page,  ...
+- Extract components that appear on various views and test them separately instead of making assertions in each test where this component is used. For example, such components could be the page header and side navigation.
+- Dedicate at least one test to assert that the view displays its UI elements in place on initial load. Assert on UI changes in tests that trigger these changes.
```
See my remark on that one.

kschiffer commented on cypress/integration/console/applications/create.spec.js:
```diff
@@ -0,0 +1,138 @@
+    cy.findDescriptionByLabelText('Network Server address')
+      .should('contain', 'Leave empty to link to the Network Server in the same cluster')
+      .and('be.visible')
+    cy.findByRole('button', { name: 'Create application' }).should('be.visible')
```
Nit: Maybe this selector config is worth an alias command `cy.findButtonByRole('Create application')`?

bafonins replied:
we might consider adding this. personally, i dont really see how we can benefit from this atm

kschiffer commented on cypress/integration/console/shared/header.spec.js:
```diff
@@ -0,0 +1,73 @@
+describe('Header', () => {
+  describe('Console logout', () => {
+    const logout = userName => {
+      cy.findByTestId('profile-dropdown')
```
More realistic selection would be something resembling

> find a dropdown containing your name in the header section

bafonins replied:
Indeed. I think it also makes sense use `cy.within` when testing shared components. Added this to the docs.

bafonins commented:
> While it is easy to pull off for simple views like Login, it becomes pretty tedious to check for all elements in views like entity overviews

There are 2 options on how solve this:
1. Tests various UI components separately. Similar to headers and side-navigation component as mentioned in the docs.
2. Do not pollute a single `it` with assertions on all UI components. There is nothing wrong in creating a bunch of `it`'s for several parts of a complex view.

> It also locks us in too much into a specific structure of the UI and its implementation  

I dont know how one can test UI without taking into account the structure of the view.

> Instead of checking UI without context, we can obtain the same validation implicitly by investing in writing more test cases, specifically test cases that are focused on the accesibility of information. E.g. for application overviews we should have test cases (based on user stories) that verify that all relevant information can be obtained. Such tests cases will also be better decoupled from implementation details.

Not sure what you mean. Can you provide an example or more details?

> We might also want to look into ways to make use of the generated screenshots, e.g. for cheap sanity-checking before making a release

Visual testing can indeed help, but I would suggest to discuss and implement this in a separate PR. This is not a simple task.

> I think it's ok to perform login and user registration programmatically, once we have tested it in a dedicated smoke test. Otherwise, we will end up with a lot of repetitive work for the test runner without gaining any value from it

Agreed. Changed this.

> I see you did not add any tests to cypress/integration/console/applications/create.spec.js that actually perform a successful submit. Just to be sure, we would plan to do some tests that test successful submitting for different scenarios as well, right? This should not be left to the smoke tests only.

Ofc, I added tests for application creation solely to showcase how we can augment the stack configuration in tests.

> Instead of creating users programmatically, why not using a seed with a default user? Considering how many tests are using authenticated views, this should speed up the tests quite a bit.

A single user is not enough. Consider, for example, testing collaborators functionality or admin functionality over the users. Additionally, in some cases we want different users to be used within the same context (`describe`) but different tests (`it`), say, for testing entity listing for various scenarios (we dont want previously created/updated entities to affect other tests. if we use a single user we have to drop the database between the tests and this takes longer than programmatically creating a new user). The question is how many users should we seed initially? Should take the approach taken in registry tests in the identity server where all necessary users are defined before running the tests? 

Also, we already seed the `admin` user. My plan is to use that user to perform specific actions like changing the state of the user, update entity rights, etc.

This is an open discussion and Im not saying that we should discuss/implement this. IMO, looking into ways to speed up tests makes sense when running e2e tests takes ~10mins. Let me know if you have a clear idea how this can be implemented within this PR addressing issues mentioned above.

> IMO, if we do integration tests, we should not limit ourselves to behavior that we can control in the frontend. The user also does not care from which component a usability issue originates from. If we write our end-to-end tests based on real-life scenarios, we should not make different assertions based on what implementation details are present in the background, just like the user wouldn't do that too.

1. We are writing end-to-end tests, not integration tests.
2. What exactly are you testing when asserting on text in the error notification for the error returned from the backend?
	- Does the backend validate requests? Thats the responsibility of the backend to test this behavior.
	- Does the backend return errors for a particular request of a certain type? Thats the responsibility of the backend to test this behavior.
	- Is the `<ErrorNotification />` component working and not breaking the view? Well, thats already covered when the component is rendered. Ideally, we should integrate cypress into storybook and run unit/integration tests for each component.
	- Does the `<ErrorNotification />` component render the error message? 1-2 tests are enough for this and this can be achieved with simply unit testing.
	- Does the `<ErrorNotification />` component render `An unknown error occurred. Please try again later.` for unknown errors? Same here.
3. I dont think that asserting on a particular _text_ gives us much more confidence than asserting on functional requirements:
 	- that the notification is shown and visible when form submission failed.
	- that the url stays the same
	- that the form state is preserved
4. Overall, I dont think that asserting on any hardcoded text is a good idea, but taking into account the number of issues we had with missing or incorrect frontend defined messages we should assert on these. We dont have the tooling to verify that the message exists if someone references `sharedMessages.someMessage` (btw typescript can solve this issue), thats why have I added support for asserting on field errors, warnings, etc.
5. In case of testing error responses, how many assertions is enough? Should we test all permutations of inputs that passes the frontend defined validation?

I suggest to find out what works better in our case. Lets define a bunch of assertions on backend returned data and experiment for a while to see how to proceed further. At the end of the day all the guidelines are just suggestions such that anyone from the dev team can quickly start writing tests that bring as much value as possible with less investment.

kschiffer commented on pkg/webui/template.go:
```diff
@@ -89,17 +89,25 @@ const appHTML = `
+				SITE_TITLE:{{.Title}},
+				SITE_SUB_TITLE:{{.SubTitle}},
+				SENTRY_DSN:{{.SentryDSN}},
+				{{with .CSRFToken}}CSRF_TOKEN:{{.}}{{end}},
```
What I mean is the line has an ending comma while being in a conditional statement of the template. When there is no `PAGE_DATA`, it could result in syntax errors since there will be a trailing comma.
A solution would be to not use these conditional statements at the end of the object assignment.

kschiffer commented on tools/mage/dev.go:
```diff
@@ -155,6 +241,15 @@ func (Dev) InitStack() error {
 	)
 }
 
+// StartDevStack starts TTS in end-to-end test configuration.
```
`in development configuration`

kschiffer replied:
So actually, I would like to retain the option to run e2e tests on a separate test database. Right now, running the tests will drop the development database. If I develop, I want to keep my current DB setup independently from the e2e tests, which is why I designed it to be separated in my PR.

kschiffer commented on cypress/integration/smoke/registration/index.js:
```diff
@@ -0,0 +1,78 @@
+    .should('be.visible')
+    .should('contain', 'You have successfully registered and can login now')
+
+  // Login
```
```suggestion
  // Login.
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,170 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+- Select DOM elements using text captions and labels when possible.
+  - Select form fields by its label via `cy.findByLabelText`, e.g. `cy.findByLabelText('User ID')`. Same for field errors,warnings and descriptions, use `cy.findErrorByLabelText`, `cy.findWarningByLabelText` and `cy.findDescriptionByLabelText`.
+  - Select buttons, links, tabs and other elements that are described by [ARIA roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#Roles) via `cy.findByRole`, e.g. `cy.findByRole('button', {name: 'Submit'})`.
+  - Select text elements via `cy.findByText`, e.g. `cy.findByText('Change password', {selector: 'h1'})`.
```
I would actually not encourage to make a selection (partially) based on the element type. It would be nicer if it was possible to select headings in general `h1, h2, h3, h4, h5`. Ultimately, what matters is that it is a heading, not exactly what kind of heading.

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,170 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+2. `pkg/cypress/{app}/shared` contains all test specification not directly related to a single entity or a view. For example, `side-navigation.spec.js` or `header.spec.js` must be placed into the `cypress/console/shared` folder because both components are present on multiple views and are partially related to the stack entities. Make sure to scope cypress selections within the tested component us ...
+3. `pkg/cypress/smoke` contains tests that simulate a complete user story trying to do almost everything a typical user would do. For example, a typical smoke test can verify that the user is able to register, login, create application and register The Things Uno. For more details and diffeence between regular end-to-end and smoke tests see the [End-to-end tests structure](#organizing-end-to-end- ...
+4. and 5. `Cypress` stores screenshots and videos to the appropriate folder after running end-to-end tests. These should not be added to the repository.
+
```
```suggestion
```

kschiffer commented on cypress/fixtures/example.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
+  "body": "Fixtures are a great way to mock data for responses to routes"
```
Should we use this to fix some common user types?

bafonins replied:
What do you mean?

kschiffer replied:
Sorry, disregard this.

kschiffer commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,170 @@
+  })
+})
+
+// Selectors
```
```suggestion
// Selectors.
```

kschiffer commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,170 @@
+Cypress.Commands.add('findWarningByLabelText', label => {
+  getFieldDescriptorByLabel(label).as('warning')
+
+  // Check for the warning icon
```
```suggestion
  // Check for the warning icon.
```

kschiffer commented on package.json:
```diff
@@ -75,6 +81,7 @@
   "dependencies": {
     "@formatjs/intl-relativetimeformat": "^2.8.2",
     "@sentry/browser": "^5.19.0",
+    "@testing-library/cypress": "^6.0.0",
```
This should be a `devDependency`, I guess?

kschiffer commented on pkg/webui/oauth/views/create-account/index.js:
```diff
@@ -199,19 +199,22 @@ export default class CreateAccount extends React.PureComponent {
                 title={sharedMessages.userId}
                 name="user_id"
                 autoComplete="username"
+                data-test-id="user_id"
```
These should not be necessary anymore.

kschiffer commented on pkg/webui/oauth/views/login/index.js:
```diff
@@ -147,8 +147,8 @@ export default class OAuth extends React.PureComponent {
                 title={sharedMessages.password}
-                component={Input}
                 name="password"
+                component={Input}
```
Remove this diff.

kschiffer commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+          name: Frontend test screenshots
+          path: cypress/screenshots
+        if: ${{ failure() }}
+      - name: Upload videos if failed
```
Should we also upload the coverage report?

htdvisser commented on tools/mage/general.go:
```diff
@@ -33,3 +34,11 @@ func targetError(err error) error {
 	return nil
 }
+
+func pathExists(filename string) bool {
+	_, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
```
This is technically incorrect. If you don't have permission to read the dir in which you're looking for the file, it would return that the file exists, while you don't actually know. 

Future developers who see that there is a `pathExists` in this package might use it incorrectly, so I don't think it's a good idea to do it this way.

Also, you can shorten L40-43 to `return !os.IsNotExist(err)`, which makes me wonder if it's really worth that one line you might save in the places where you'd use this.

htdvisser commented on DEVELOPMENT.md:
````diff
@@ -716,10 +716,169 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+> Note: Currently, we test our frontend only in Chromium based browsers.
+
+You can run `Cypress` in the headless mode by running the following command:
+```bash
````
Newlines between blocks

htdvisser commented on DEVELOPMENT.md:
````diff
@@ -716,10 +716,169 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+$ tools/bin/mage endToEnd:CypressHeadless
+```
+
+You can run `Cypress` in the interactive mode by running the following command:
+```bash
````
Newlines between blocks

htdvisser commented on DEVELOPMENT.md:
````diff
@@ -716,10 +716,169 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+  </div>
+  ```
+
+  ```js
+    // Bad. This assertions will pass while not being visible to the user.
+    cy.findByTestId('test').should('exists')
+
+    // Good. This assertion will rightfully fail.
+    cy.findByTestId('test').should('be.visible')
+  ```
````
Strange indentation inside the code block?

htdvisser commented on DEVELOPMENT.md:
````diff
@@ -716,10 +716,169 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+
+- Global text summary.
+```bash
+  npx nyc report --reporter=text-summary
````
No $? Also below

htdvisser commented on tools/mage/dev.go:
```diff
@@ -70,10 +76,90 @@ func execDockerCompose(args ...string) error {
+
+// SQLMakeSnapshot stores the current cockroach data folder for later restores.
+func (Dev) SQLMakeSnapshot() error {
+	if mg.Verbose() {
```
Shouldn't we first make sure the database is stopped?

htdvisser commented on tools/mage/dev.go:
```diff
@@ -70,10 +76,90 @@ func execDockerCompose(args ...string) error {
+
+// DBDump performs an SQL database dump of the dev database to the .cache folder.
+func (Dev) DBDump() error {
+	if mg.Verbose() {
```
Shouldn't we first make sure the database is running?

bafonins replied:
Shall we do that? The main use of `DBDump` is to get the database dump after `InitStack` while `InitStack` does not depend on `DBStart`.
Also, if the database is not started a proper error is displayed.

htdvisser commented on tools/mage/dev.go:
```diff
@@ -70,10 +76,90 @@ func execDockerCompose(args ...string) error {
+
+// DBRestore restores the dev database using a previously generated dump.
+func (Dev) DBRestore() error {
+	mg.Deps(Js.Deps)
```
Shouldn't we first make sure the database is running?

htdvisser commented on tools/mage/scripts/restore-db-dump.js:
```diff
@@ -0,0 +1,55 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
```
I'm not a big fan of using mixed Mage / JavaScript for our dev tooling here.

htdvisser commented on tools/mage/dev.go:
```diff
@@ -70,10 +76,90 @@ func execDockerCompose(args ...string) error {
+	if err := os.MkdirAll(".cache", 0755); err != nil {
+		return err
+	}
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", devDatabaseName, "--insecure")
```
Not a big fan of this, as it assumes we always use CockroachDB (which we don't; in CI we use PostgreSQL)

bafonins replied:
But this actually works in CI 🤷‍♂️

rvolosatovs replied:
Split line for readability - see e.g. https://github.com/TheThingsNetwork/lorawan-stack/blob/31119a0db8e9a7933e021195623f50ba5420f2df/tools/mage/dev.go#L135-L140

htdvisser commented on tools/mage/dev.go:
```diff
@@ -70,10 +76,90 @@ func execDockerCompose(args ...string) error {
+	if mg.Verbose() {
+		fmt.Println("Restoring database from dump")
+	}
+	return sh.Run("node", filepath.Join("tools", "mage", "scripts", "restore-db-dump.js"), "--db", devDatabaseName, "--dump", filepath.Join(".cache", "sqldump.sql"))
```
Not really happy with an external script that just runs

```
DROP DATABASE <NAME>
CREATE DATABASE <NAME>
```

and then the queries in `sqldump.sql`.

We can easily do that here directly.

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+  pull_request:
+    paths:
+      - '*.go'
+      - '.revive.toml'
```
This should not trigger end-to-end tests

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+name: End-to-end Tests
+on:
+  pull_request:
+    paths:
```
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764/files#r447485316

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+    name: Frontend based (cypress)
+    runs-on: ubuntu-18.04
+    env:
+      should_cache: true # Might be useful for debugging.
```
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764/files#r447485665

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -55,12 +59,16 @@ func (Dev) Misspell() error {
+	databaseURI           = fmt.Sprintf("postgresql://root@localhost:26257/%s?sslmode=disable", devDatabaseName)
 )
 
+var devDatabases = []string{sqlDatabase, redisDatabase}
```
nit; this can be in the same `var` block above

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -70,10 +78,97 @@ func execDockerCompose(args ...string) error {
+	}
+	message := []byte(output)
+	err = ioutil.WriteFile(filepath, message, 0644)
+	return err
```
```suggestion
	return ioutil.WriteFile(filepath, []byte(output), 0644)
```

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -70,10 +78,97 @@ func execDockerCompose(args ...string) error {
 	return err
 }
 
+func execDockerComposeWithOutput(filepath string, args ...string) error {
```
This function is way too specific and is inconsistent with Go analogues:
https://github.com/TheThingsNetwork/lorawan-stack/blob/31119a0db8e9a7933e021195623f50ba5420f2df/tools/mage/go.go#L54-L72

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -70,10 +78,97 @@ func execDockerCompose(args ...string) error {
+	if err = db.Exec(string(b)).Error; err != nil {
+		return err
+	}
+
```
nit; remove newline

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -70,10 +78,97 @@ func execDockerCompose(args ...string) error {
+		return nil
+	}
+
+	if err = db.Exec(fmt.Sprintf("DROP DATABASE %s;", devDatabaseName)).Error; err != nil {
+		return err
+	}
+	if err = db.Exec(fmt.Sprintf("CREATE DATABASE %s;", devDatabaseName)).Error; err != nil {
+		return err
+	}
+	if err = db.Exec(string(b)).Error; err != nil {
+		return err
+	}
+
+	return nil
```
```suggestion
	return db.Exec(fmt.Sprintf(`DROP DATABASE %s;
CREATE DATABASE %s;
%s`,
 devDatabaseName, devDatabaseName, string(b)),
).Error
```

rvolosatovs commented on tools/mage/go.go:
```diff
@@ -71,6 +71,10 @@ func outputGo(cmd string, args ...string) (string, error) {
 	return buf.String(), nil
 }
 
+func runGoMute(args ...string) error {
```
This is only ever used once in the whole codebase, remove and just use `execGo(nil, os.Stderr, "run", args...)` explicitly

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -155,6 +250,15 @@ func (Dev) InitStack() error {
+	if mg.Verbose() {
+		fmt.Println("Starting the Stack")
+	}
+	return runGoMute("./cmd/ttn-lw-stack", "start")
```
Use `filepath.Join`

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -76,7 +77,32 @@ func (js Js) runWebpack(config string, args ...string) error {
+	if err != nil {
+		return err
+	}
+	return js.runYarnCommand("wait-on", append([]string{
```
`append` and `[]string{}` should not be necessary here

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -76,7 +77,32 @@ func (js Js) runWebpack(config string, args ...string) error {
+func (js Js) runCypress(command string, args ...string) error {
+	mg.Deps(js.waitOn)
+
+	return js.runYarnCommand("cypress", append([]string{command, fmt.Sprintf("--config-file=%s", filepath.Join("config", "cypress.json")), "--config", fmt.Sprintf("baseUrl=%s", js.getURL())}, args...)...)
```
split line

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -76,7 +77,32 @@ func (js Js) runWebpack(config string, args ...string) error {
+
+func (js Js) runCypress(command string, args ...string) error {
+	mg.Deps(js.waitOn)
+
```
nit; remove newline

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -76,7 +77,32 @@ func (js Js) runWebpack(config string, args ...string) error {
+	return js.runYarnCommand("cypress", append([]string{command, fmt.Sprintf("--config-file=%s", filepath.Join("config", "cypress.json")), "--config", fmt.Sprintf("baseUrl=%s", js.getURL())}, args...)...)
+}
+
+func (js Js) getURL() string {
```
1. Avoid `getX()`, just write `X()` - this is a convention in Go
2. `URL` is too ambiguous
```suggestion
func (js Js) frontendURL() string {
```
Or maybe you have a better suggestion?

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -76,7 +77,32 @@ func (js Js) runWebpack(config string, args ...string) error {
+
+func (js Js) getURL() string {
+	if js.isProductionMode() {
+		return "http://localhost:1885"
```
Would be nice to at least have the ports defined as consts above

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -70,10 +78,97 @@ func execDockerCompose(args ...string) error {
+	if err := os.MkdirAll(".cache", 0755); err != nil {
+		return err
+	}
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", devDatabaseName, "--insecure")
```
Moreover, this is the only usage of `execDockerComposeWithOutput` in the whole codebase - just remove the helper and use `execDockerCompose` explicitly here

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -76,7 +82,36 @@ func (js Js) runWebpack(config string, args ...string) error {
+
+func (js Js) runCypress(command string, args ...string) error {
+	mg.Deps(js.waitOn)
+	// return js.runYarnCommand("cypress", append([]string{command, fmt.Sprintf("--config-file=%s", filepath.Join("config", "cypress.json")), "--config", fmt.Sprintf("baseUrl=%s", js.frontendURL())}, args...)...)
```
no commented-out code please

kschiffer commented on cypress/integration/oauth/change-password.spec.js:
```diff
@@ -0,0 +1,126 @@
+  })
+
+  // TODO: Remove `skip` after https://github.com/TheThingsNetwork/lorawan-stack/issues/2951.
+  it.skip('succeeds changing password when revoking access', () => {
```
```suggestion
  it('succeeds changing password when revoking access', () => {
```

Issue is resovled.

kschiffer commented on cypress/integration/oauth/change-password.spec.js:
```diff
@@ -0,0 +1,126 @@
+  })
+
+  // TODO: Remove `skip` after https://github.com/TheThingsNetwork/lorawan-stack/issues/2951.
+  it.skip('succeeds changing password without revoking access', () => {
```
Same.

kschiffer commented on cypress/integration/console/applications/create.spec.js:
```diff
@@ -0,0 +1,134 @@
+      .and('eq', 'my-new-application')
+    cy.findByLabelText('Application name')
+      .should('be.visible')
+      .should('be.visible')
```
```suggestion
```

kschiffer commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,175 @@
+// Helper function to quickly seed the database to a fresh state using a
+// previously generated sql dump.
+Cypress.Commands.add('dropAndSeedDatabase', () => {
+  // Cy.exec('node tools/mage/scripts/restore-db-dump.js')
```
```suggestion
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,176 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+  - For example, when testing registration of The Things Uno:
+    1. Add test file `cypress/integration/smoke/devices/create.js`
+    2. Describe the whole user story to register the device including creating an application (or using an existing one), link the application and create the end device.
+- For example, when testing registration of The Things Uno:
```
```suggestion
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,176 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+  - Each test must follow the following naming: `{context}.spec.js`.
+  - One test file must have end-to-end tests dedicated only to a specific entity or view.
+2. `pkg/cypress/{app}/shared` contains all test specification not directly related to a single entity or a view. For example, `side-navigation.spec.js` or `header.spec.js` must be placed into the `cypress/console/shared` folder because both components are present on multiple views and are partially related to the stack entities. Make sure to scope cypress selections within the tested component us ...
+3. `pkg/cypress/smoke` contains tests that simulate a complete user story trying to do almost everything a typical user would do. For example, a typical smoke test can verify that the user is able to register, login, create application and register The Things Uno. For more details and diffeence between regular end-to-end and smoke tests see the [End-to-end tests structure](#organizing-end-to-end- ...
```
```suggestion
3. `pkg/cypress/integration/smoke` contains tests that simulate a complete user story trying to do almost everything a typical user would do. For example, a typical smoke test can verify that the user is able to register, login, create application and register The Things Uno. For more details and diffeence between regular end-to-end and smoke tests see the [End-to-end tests structure](#organizing-end-to-end-tests) section.
```

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,176 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+
+##### Smoke tests
+
+We distinguish between regular end-to-end tests and smoke tests. With smoke tests we try to accumulate application state (redux store, local storage, application cache, etc.) and verify that the application works as expected. When writing smoke tests we comply with he following guidelines:
```
```suggestion
We distinguish between regular end-to-end tests and smoke tests. While regular end-to-end tests are scoped to a specific view or component and tests those in depth, smoke tests are testing complete user stories that are critical to the overall initegrity of the application and usually comprise multiple components and views, e.g. login flow, user registration or creation of applications. When writing smoke tests we comply with he following guidelines:
```

Clarifying the difference between regular e2e and smoke tests a little more.

kschiffer commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,176 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+We distinguish between regular end-to-end tests and smoke tests. With smoke tests we try to accumulate application state (redux store, local storage, application cache, etc.) and verify that the application works as expected. When writing smoke tests we comply with he following guidelines:
+
+- Smoke tests are testing complete user stories in a **wide and shallow** manner, meaning:
+  - performing some complex and critical flow that touches multiple components, API's and/or views
```
```suggestion
  - performing some complex and critical flow that touches multiple components, APIs and/or views
```

rvolosatovs commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,81 @@
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+      - name: Initialize messages and locales cache
```
```suggestion
      - name: Initialize message and locale cache
```

rvolosatovs commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,81 @@
+        with:
+          path: public
+          key: ${{ runner.os }}-public-cache-${{ hashFiles('pkg/webui/**') }}-${{ hashFiles('sdk/js/**/*.js') }}-${{ hashFiles('config/webpack.config.babel.js') }}-${{ hashFiles('**/yarn.lock') }}-${{ hashFiles('pkg/ttnpb/field_mask_validation.go') }}
+      - name: Initialize JS SDK cache
```
Why is this only performed here and not in other workflows using JS SDK?
Please add to all relevant workflows in order to avoid inconsistencies. (can be in a separate PR)

rvolosatovs replied:
Is there any action taken in response to this?
Are you going to update this PR? Are you going to open another PR before merging this one? Are you going to file an issue?

bafonins replied:
I removed caching steps that do no appear in other js workflow. We can add those later if we experience any noticeable performance issues.

kschiffer replied:
Note that I was able to bring the run time down [to about 4 minutes](https://github.com/kschiffer/lorawan-stack/runs/783055521) under optimal circumstances with the caching I had applied in the other e2e PR.

rvolosatovs commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,81 @@
+            .cache/messages
+            pkg/webui/console/locales/.backend
+          key: ${{ runner.os }}-messages-cache-${{ hashFiles('pkg/webui/**/*.js') }}-${{ hashFiles('config/messages.json')}}
+      - name: Initialize public folder cache
```
IDEM

rvolosatovs commented on DEVELOPMENT.md:
```diff
@@ -716,10 +716,175 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
 
+### Unit Tests
+
+To run unit tests, observe the following mage targets:
```
```suggestion
To run unit tests, use the following mage targets:
```
or *run*

rvolosatovs commented on DEVELOPMENT.md:
````diff
@@ -716,10 +716,175 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+$ tools/bin/mage endToEnd:CypressInteractive
+```
+
+#### Running frontend end-to-end tests in CI
````
Not sure what this section is for, we can just read the workflow source, right?

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -155,6 +239,15 @@ func (Dev) InitStack() error {
 
+// StartDevStack starts TTS in end-to-end test configuration.
+func (Dev) StartDevStack() error {
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
```
I would expect a running DB to be a dependency of this.
If the reason why it's not is that DB may be started externally (i.e. not via this mage file), how about a target, which would check if DB is available, and, if not, starts the DB?
